diff --git a/riscv/arith.cc b/riscv/arith.cc
new file mode 100644
index 00000000..32ce9d92
--- /dev/null
+++ b/riscv/arith.cc
@@ -0,0 +1 @@
+#include "arith.h"
diff --git a/riscv/decode.h b/riscv/decode.h
index f36c04e5..734584b3 100644
--- a/riscv/decode.h
+++ b/riscv/decode.h
@@ -79,6 +79,11 @@ public:
   insn_t(insn_bits_t bits) : b(bits) {}
   insn_bits_t bits() { return b; }
   int length() { return insn_length(b); }
+  // ++ mpise
+  uint64_t funct7() { return x(25, 7); }
+  uint64_t funct3() { return x(12, 3); }
+  uint64_t funct2() { return x(25, 2); }
+  // -- mpise
   int64_t i_imm() { return xs(20, 12); }
   int64_t shamt() { return x(20, 6); }
   int64_t s_imm() { return x(7, 5) + (xs(25, 7) << 5); }
diff --git a/riscv/encoding.h b/riscv/encoding.h
index 675b4f64..62af1951 100644
--- a/riscv/encoding.h
+++ b/riscv/encoding.h
@@ -2418,7 +2418,45 @@
 #define MASK_XPERM4 0xfe00707f
 #define MATCH_XPERM8 0x28004033
 #define MASK_XPERM8 0xfe00707f
-
+// ++ mpise
+#define MATCH_SRAIADD 0x000100B
+#define  MASK_SRAIADD 0x000707F
+#if   defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 0 )
+// !destructive
+#define MATCH_CACC    0x000000B
+#define  MASK_CACC    0x000707F
+#if   defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 0 )
+// !destructive && !stateless
+#define MATCH_MACCLU  0x000002B
+#define  MASK_MACCLU  0x000107F
+#define MATCH_MACCHU  0x000102B
+#define  MASK_MACCHU  0x000107F
+#elif defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 1 )
+// !destructive &&  stateless
+#define MATCH_MACCLU  0x000002B
+#define  MASK_MACCLU  0x000107F
+#define MATCH_MACCHU  0x000102B
+#define  MASK_MACCHU  0x000107F
+#endif
+#elif defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 1 )
+//  destructive
+#define MATCH_CACC    0x000000B
+#define  MASK_CACC    0x000707F
+#if   defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 0 )
+//  destructive && !stateless
+#define MATCH_MACCLU  0x000002B
+#define  MASK_MACCLU  0x000107F
+#define MATCH_MACCHU  0x000102B
+#define  MASK_MACCHU  0x000107F
+#elif defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 1 )
+//  destructive &&  stateless
+#define MATCH_MACCLU  0x000002B
+#define  MASK_MACCLU  0x000107F
+#define MATCH_MACCHU  0x000102B
+#define  MASK_MACCHU  0x000107F
+#endif
+#endif
+// -- mpise
 #define CSR_FFLAGS 0x1
 #define CSR_FRM 0x2
 #define CSR_FCSR 0x3
@@ -2875,7 +2913,9 @@
 #define CSR_MHPMCOUNTER29H 0xb9d
 #define CSR_MHPMCOUNTER30H 0xb9e
 #define CSR_MHPMCOUNTER31H 0xb9f
-
+// ++ mpise
+#define CSR_MACC           0x800
+// -- mpise
 #define CAUSE_MISALIGNED_FETCH 0x0
 #define CAUSE_FETCH_ACCESS 0x1
 #define CAUSE_ILLEGAL_INSTRUCTION 0x2
@@ -3961,6 +4001,12 @@ DECLARE_INSN(xperm16, MATCH_XPERM16, MASK_XPERM16)
 DECLARE_INSN(xperm32, MATCH_XPERM32, MASK_XPERM32)
 DECLARE_INSN(xperm4, MATCH_XPERM4, MASK_XPERM4)
 DECLARE_INSN(xperm8, MATCH_XPERM8, MASK_XPERM8)
+// ++ mpise
+DECLARE_INSN(cacc,    MATCH_CACC,    MASK_CACC)
+DECLARE_INSN(sraiadd, MATCH_SRAIADD, MASK_SRAIADD)
+DECLARE_INSN(macclu,  MATCH_MACCLU,  MASK_MACCLU)
+DECLARE_INSN(macchu,  MATCH_MACCHU,  MASK_MACCHU)
+// -- mpise
 #endif
 #ifdef DECLARE_CSR
 DECLARE_CSR(fflags, CSR_FFLAGS)
@@ -4419,6 +4465,9 @@ DECLARE_CSR(mhpmcounter28h, CSR_MHPMCOUNTER28H)
 DECLARE_CSR(mhpmcounter29h, CSR_MHPMCOUNTER29H)
 DECLARE_CSR(mhpmcounter30h, CSR_MHPMCOUNTER30H)
 DECLARE_CSR(mhpmcounter31h, CSR_MHPMCOUNTER31H)
+// ++ mpise
+DECLARE_CSR(macc, CSR_MACC)
+// -- mpise
 #endif
 #ifdef DECLARE_CAUSE
 DECLARE_CAUSE("misaligned fetch", CAUSE_MISALIGNED_FETCH)
diff --git a/riscv/insns/cacc.h b/riscv/insns/cacc.h
new file mode 100644
index 00000000..6fa325a4
--- /dev/null
+++ b/riscv/insns/cacc.h
@@ -0,0 +1,22 @@
+if( ( xlen == 32 ) || ( xlen == 64 ) ) {
+  int w = xlen;
+
+  #if   defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 0 )
+  // !destructive
+  uint128_t x = READ_REG( insn.rs1() );
+  uint128_t y = READ_REG( insn.rs2() );
+  uint128_t z = READ_REG( insn.rs3() );
+  #elif defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 1 )
+  //  destructive
+  uint128_t x = READ_REG( insn.rs1() );
+  uint128_t y = READ_REG( insn.rs2() );
+  uint128_t z = READ_REG( insn.rd()  );
+  #endif
+
+  uint128_t r = ( ( x + y ) >> w ) + z;
+
+  WRITE_REG( insn.rd(), r );
+}
+else {
+  abort();
+}
diff --git a/riscv/insns/macchu.h b/riscv/insns/macchu.h
new file mode 100644
index 00000000..d02fa05e
--- /dev/null
+++ b/riscv/insns/macchu.h
@@ -0,0 +1,49 @@
+if( ( xlen == 32 ) || ( xlen == 64 ) ) {
+  int w = xlen;
+
+  #if   defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 0 )
+  // !destructive
+  uint128_t x   = READ_REG( insn.rs1() );
+  uint128_t y   = READ_REG( insn.rs2() );
+  uint128_t z   = READ_REG( insn.rs3() );
+  #elif defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 1 )
+  //  destructive
+  uint128_t x   = READ_REG( insn.rs1() );
+  uint128_t y   = READ_REG( insn.rs2() );
+  uint128_t z   = READ_REG( insn.rd()  );
+  #endif
+
+  #if   defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 0 )
+  // !stateless
+  int c = p->get_csr( 0x800, insn, false ) & 0xF;
+  #elif defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 1 )
+  //  stateless
+  int c = ( ( insn.funct2() & 0x3 ) << 2 ) |
+          ( ( insn.funct3() & 0x6 ) >> 1 ) ;
+  #endif
+
+  int f_3, f_4, f_5;
+
+  if     ( c == 0xF ) { //    full-radix
+    f_3 =        1;
+    f_4 =  w      ;
+    f_5 =  w      ;
+  }
+  else if( c != 0xF ) { // reduced-radix
+    f_3 =        0;
+    f_4 =  w      ;
+    f_5 = (w-15)+c;
+  }
+
+  uint128_t t_0 = x * y;
+  uint128_t t_1 = f_3 ? z : 0;
+  uint128_t t_2 = f_3 ? 0 : z;
+
+  uint128_t m   = ( ( uint128_t )( 1 ) << f_4 ) - 1;
+  uint128_t r   = ( ( ( t_0 + t_1 ) >> f_5 ) & m ) + t_2;
+
+  WRITE_REG( insn.rd(), r );
+}
+else {
+  abort();
+}
diff --git a/riscv/insns/macclu.h b/riscv/insns/macclu.h
new file mode 100644
index 00000000..4382feed
--- /dev/null
+++ b/riscv/insns/macclu.h
@@ -0,0 +1,49 @@
+if( ( xlen == 32 ) || ( xlen == 64 ) ) {
+  int w = xlen;
+
+  #if   defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 0 )
+  // !destructive
+  uint128_t x   = READ_REG( insn.rs1() );
+  uint128_t y   = READ_REG( insn.rs2() );
+  uint128_t z   = READ_REG( insn.rs3() );
+  #elif defined( MPISE_DESTRUCTIVE ) && ( MPISE_DESTRUCTIVE == 1 )
+  //  destructive
+  uint128_t x   = READ_REG( insn.rs1() );
+  uint128_t y   = READ_REG( insn.rs2() );
+  uint128_t z   = READ_REG( insn.rd()  );
+  #endif
+
+  #if   defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 0 )
+  // !stateless
+  int c = p->get_csr( 0x800, insn, false ) & 0xF;
+  #elif defined( MPISE_STATELESS   ) && ( MPISE_STATELESS   == 1 )
+  //  stateless
+  int c = ( ( insn.funct2() & 0x3 ) << 2 ) |
+          ( ( insn.funct3() & 0x6 ) >> 1 ) ;
+  #endif
+
+  int f_0, f_1, f_2;
+
+  if     ( c == 0xF ) { //    full-radix
+    f_0 =        1;
+    f_1 =  w      ;
+    f_2 =        0;
+  }
+  else if( c != 0xF ) { // reduced-radix
+    f_0 =        0;
+    f_1 = (w-15)+c;
+    f_2 =        0;
+  }
+
+  uint128_t t_0 = x * y;
+  uint128_t t_1 = f_0 ? z : 0;
+  uint128_t t_2 = f_0 ? 0 : z;
+
+  uint128_t m   = ( ( uint128_t )( 1 ) << f_1 ) - 1;
+  uint128_t r   = ( ( ( t_0 + t_1 ) >> f_2 ) & m ) + t_2;
+
+  WRITE_REG( insn.rd(), r );
+}
+else {
+  abort();
+}
diff --git a/riscv/insns/sraiadd.h b/riscv/insns/sraiadd.h
new file mode 100644
index 00000000..510c93a4
--- /dev/null
+++ b/riscv/insns/sraiadd.h
@@ -0,0 +1,14 @@
+if( ( xlen == 32 ) || ( xlen == 64 ) ) {
+  int w = xlen;
+
+   int64_t x   = READ_REG( insn.rs1() );
+   int64_t y   = READ_REG( insn.rs2() );
+  uint64_t imm = insn.funct7() & 0x3F;
+
+   int64_t r   = x + ( y >> imm );
+
+  WRITE_REG( insn.rd(), r );
+}
+else {
+  abort();
+}
diff --git a/riscv/riscv.mk.in b/riscv/riscv.mk.in
index bc512bfa..ba79a54a 100644
--- a/riscv/riscv.mk.in
+++ b/riscv/riscv.mk.in
@@ -49,6 +49,7 @@ riscv_precompiled_hdrs = \
 	insn_template.h \
 
 riscv_srcs = \
+	arith.cc \
 	processor.cc \
 	execute.cc \
 	dts.cc \
@@ -1083,6 +1084,10 @@ riscv_insn_ext_zicfiss = \
 	c_sspush_x1 \
 	c_sspopchk_x5 \
 
+# ++ mpise
+riscv_insn_mpise = cacc sraiadd macchu macclu
+# -- mpise
+
 riscv_insn_ext_zvk = \
 	$(riscv_insn_ext_zvbb) \
 	$(riscv_insn_ext_zvbc) \
@@ -1093,6 +1098,7 @@ riscv_insn_ext_zvk = \
 	$(riscv_insn_ext_zvksh) \
 
 riscv_insn_list = \
+	$(riscv_insn_mpise) \
 	$(riscv_insn_ext_i) \
 	$(riscv_insn_ext_c) \
 	$(riscv_insn_ext_f) \
